dhcp-protocol/src/main/java/org/anarres/dhcp/v6/io/Dhcp6MessageDecoder.java

```diff
@@ -4,32 +4,133 @@
+        LOG.debug("Message type: {}", msgType);
+
+
+        Dhcp6Message message;
+        if(isRelayedMessage(msgType)) {
+            LOG.debug("Message relay");
+
+            message = new Dhcp6RelayMessage();
+            final Dhcp6RelayMessage relayMsg = ((Dhcp6RelayMessage) message);
+
+            relayMsg.setHopCount(getByte(buffer));
+            LOG.debug("Hop count: {}", message.getTransactionId());
+
+            relayMsg.setLinkAddress(getInetAddress(buffer, (byte) 16));
+            LOG.debug("Link address: {}", message.getTransactionId());
+
+            relayMsg.setPeerAddress(getInetAddress(buffer, (byte) 16));
+            LOG.debug("Peer address: {}", message.getTransactionId());
+        } else {
+            message = new Dhcp6Message();
+            message.setMessageType(msgType);
 
-        Dhcp6Options options = decodeOptions(buffer);
-        message.setOptions(options);
+            message.setTransactionId(getTxId(buffer));
+            LOG.debug("Transaction ID: {}", message.getTransactionId());
 
+            Dhcp6Options options = decodeOptions(buffer);
+            message.setOptions(options);
+        }
+
+        LOG.debug("Dhcp6 message decoded: {}", message);
+        message.setMessageType(msgType);
         return message;
+
+    }
+
+    private InetAddress getInetAddress(final ByteBuffer bufferm, final byte size) throws Dhcp6Exception {
+        Preconditions.checkArgument(bufferm.remaining() >= size);
+        final byte[] ipBytes = new byte[size];
+        bufferm.get(ipBytes, bufferm.position(), size);
+        try {
+            return InetAddress.getByAddress(ipBytes);
+        } catch (UnknownHostException e) {
+            throw new Dhcp6Exception("Unable to parse inet address from: " + BaseEncoding.base16().encode(ipBytes), e);
+        }
+    }
+
+    static boolean isRelayedMessage(final Dhcp6MessageType type) {
+        return type == Dhcp6MessageType.DHCP_RELAY_FORW || type == Dhcp6MessageType.DHCP_RELAY_REPL;
+    }
+
+    /**
+     * https://tools.ietf.org/html/rfc3315#section-15.1
+     */
+    private static int getTxId(final ByteBuffer buffer) {
+        Preconditions.checkArgument(buffer.remaining() > 3);
+        final byte[] txBytes = new byte[4];
+        buffer.get(txBytes, 1, 3);
+        txBytes[0] = 0;
+        return Ints.fromByteArray(txBytes);
+    }
+
+    /**
+     * https://tools.ietf.org/html/rfc3315#section-5.3
+     */
+    @Nonnull
+    private static Dhcp6MessageType getMsgType(final ByteBuffer buffer) throws Dhcp6Exception.UnknownMsgException {
+        final byte type = getByte(buffer);
+        final Dhcp6MessageType dhcp6MessageType = Dhcp6MessageType.forTypeCode(type);
+        Dhcp6Exception.UnknownMsgException.check(dhcp6MessageType, type);
+        return dhcp6MessageType;
+    }
+
+    private static byte getByte(final ByteBuffer buffer) {
+        Preconditions.checkArgument(buffer.remaining() >= 1);
```

Comment:
Is the IllegalArgumentException more illuminating than the BufferUnderFlowException which would be thrown by the unguarded get() call?